### PR TITLE
MOD-4551: Workaround indexing u64 integer greater than i64::MAX

### PR DESCRIPTION
**Pull request by oshadmi**
Currently there is an API `get_long(&self) -> i64`
Workaround to avoid panic on unwrap when an integer value greater than `i64::MAX` is indexed by RediSearch.
Workaround is needed until an API `get_ulong(&self) -> u64` is added to `SelectValue` trait.

Related #875

### CODE REVIEW

**Review by MeirShpilraien — DISMISSED** [review]
Maybe we should document it?

**Comment by oshadmi** [comment]
> Maybe we should document it?

Sure.
@nermiller Should we document this workaround and limitation in the release notes, in the documentation itself or both?

**Comment by nermiller** [comment]
@oshadmi both would be good until the limitation is removed. Let's work on adding a note.

File: tests/pytest/test.py [hunk 1]
```diff
@@ -1280,6 +1280,87 @@ def testFilterPrecedence(env):
...
+    r.expect('JSON.SET', 'doc_int_ok', '$', json.dumps(doc_int_ok)).ok()
+    r.expect('JSON.SET', 'doc_float_ok', '$', json.dumps(doc_float_ok)).ok()
+    r.expect('JSON.SET', 'doc_float_ok_2', '$', json.dumps(doc_float_ok_2)).ok()
+
+    def check_object_values(obj, name, epsilon):
+        # Test values from JSON.GET are equal to JSON.SET
+        # Check no crash (using get_long internally)
+        arr = []
+        for k, v in iter(obj.items()):
+            r.assertTrue(True, message='GET {}={}'.format(k, v))
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:1331]: ?

File: tests/pytest/test.py [hunk 2]
```diff
@@ -1280,6 +1280,87 @@ def testFilterPrecedence(env):
...
+            res = r.execute_command('JSON.GET', name, '$.{}'.format(k))
+            r.assertAlmostEqual(json.loads(res)[0], v, epsilon, message=res)
+            try:
+                # Check no crash on overflow
+                res = r.execute_command('JSON.NUMINCRBY', name, '$.{}'.format(k), '1')
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:1336]: Can we verify the result?

File: tests/pytest/test.py [hunk 3]
```diff
@@ -1280,6 +1280,87 @@ def testFilterPrecedence(env):
...
+                #r.assertAlmostEqual(json.loads(res)[0], v+1, epsilon, message=res)
+                res = r.execute_command('JSON.NUMMULTBY', name, '$.{}'.format(k), '2')            
+                #r.assertAlmostEqual(json.loads(res)[0], v*2, epsilon, message=res)
+            except Exception as e:
+                pass
+            arr.append(v)
+        # Check no crash using values in a filter
+        r.expect('JSON.SET', 'arr', '$', json.dumps(arr)).ok()
+        res = r.execute_command('JSON.GET', 'arr', '$[?(@!=0)]')
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:1345]: Not following, shouldn't this return error now?

File: tests/pytest/test.py [hunk 4]
```diff
@@ -1280,6 +1280,87 @@ def testFilterPrecedence(env):
...
+        for res_i, arr_i in zip(json.loads(res), arr):
+            r.assertAlmostEqual(res_i, arr_i, epsilon, message=res)
+        # Check no crash with other commands 
+        r.execute_command('JSON.ARRINDEX', 'arr', '$', '10')
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:1349]: Verify reply?

File: tests/pytest/test.py [hunk 5]
```diff
@@ -1280,6 +1280,87 @@ def testFilterPrecedence(env):
...
+        r.execute_command('JSON.RESP', 'arr')
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:1350]: Verify reply

File: tests/pytest/test.py [hunk 6]
```diff
@@ -1280,6 +1280,87 @@ def testFilterPrecedence(env):
...
+            r.execute_command('JSON.CLEAR', name, '$.{}'.format(k))
+
+    check_object_values(doc_int_ok, 'doc_int_ok', 0)
+    check_object_values(doc_float_ok, 'doc_float_ok', sys.float_info.epsilon)
+    check_object_values(doc_float_ok_2, 'doc_float_ok_2', sys.float_info.epsilon)
+
+    # Not using json.dumps with out-of-range values here (would be converted to a string representation such as 'Infinity')
+    r.expect('JSON.SET', 'doc_bad_values', '$', '{}').ok()
+    for k, v in iter(doc_bad_values.items()):
+        r.assertTrue(True, message='SET {}={}'.format(k, v))
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:1361]: ?

**Review by MeirShpilraien — COMMENTED** [review]
I do not see where the new error `currently unsupported` are checked on the tests.

**Comment by MeirShpilraien** [comment]
Also lets update top comment.

File: tests/pytest/test.py (hunk 1 above)
Comment (oshadmi) [reply tests/pytest/test.py:1331]: This is just printing in order to see the current value in the loop

File: tests/pytest/test.py (hunk 2 above)
Comment (oshadmi) [reply tests/pytest/test.py:1336]: Some fail due to overflow. Could be done such as by moving values that overflow to separate keys.

File: tests/pytest/test.py (hunk 3 above)
Comment (oshadmi) [reply tests/pytest/test.py:1345]: Did not propagate the error all the way up.
A failed term (a long which is u64) is filtered out since evaluated as `NotCmparable`

File: tests/pytest/test.py (hunk 6 above)
Comment (oshadmi) [reply tests/pytest/test.py:1361]: This is just printing in order to see the current value in the loop

File: tests/pytest/test.py (hunk 1 above)
Comment (MeirShpilraien) [reply tests/pytest/test.py]: We have debugPrint for that
https://github.com/RedisLabsModules/RLTest/blob/09ced571f7d706f39c6a505d84ea2ebb05653b9a/RLTest/env.py#L58

File: tests/pytest/test.py (hunk 2 above)
Comment (MeirShpilraien) [reply tests/pytest/test.py:1345]: What do you mean fail? Do we get an error? If so, isn't it a breaking change?
And in any case we can verfit the error.

File: tests/pytest/test.py (hunk 3 above)
Comment (MeirShpilraien) [reply tests/pytest/test.py:1369]: So we need to decide if this might be considered a breaking change in the future? @gkorland ?

File: tests/pytest/test.py (hunk 6 above)
Comment (MeirShpilraien) [reply tests/pytest/test.py:1397]: Debugprint

File: src/commands.rs [hunk 7]
```diff
@@ -128,7 +128,9 @@ impl<'a, V: SelectValue + 'a> KeyValue<'a, V> {
                 }
             }
 
-            SelectValueType::Long => RedisValue::Integer(v.get_long()),
+            SelectValueType::Long => {
+                RedisValue::Integer(v.get_long().map_or_else(|_| i64::MAX, |v| v))
```
Comment (MeirShpilraien) [inline src/commands.rs:132]: Isn't this can be consider a breaking change in the future? Can we raise error?

File: src/commands.rs [hunk 8]
```diff
@@ -417,7 +419,10 @@ impl<'a, V: SelectValue + 'a> KeyValue<'a, V> {
         match (a.get_type(), b.get_type()) {
             (SelectValueType::Null, SelectValueType::Null) => true,
             (SelectValueType::Bool, SelectValueType::Bool) => a.get_bool() == b.get_bool(),
-            (SelectValueType::Long, SelectValueType::Long) => a.get_long() == b.get_long(),
+            (SelectValueType::Long, SelectValueType::Long) => match (a.get_long(), b.get_long()) {
+                (Ok(a), Ok(b)) => a == b,
+                _ => false,
```
Comment (MeirShpilraien) [inline src/commands.rs:424]: Same question here, wouldn't it will be better to raise error? @gkorland ?

File: src/commands.rs [hunk 9]
```diff
@@ -1933,7 +1938,7 @@ pub fn json_clear<M: Manager>(manager: M, ctx: &Context, args: Vec<RedisString>)
 
     let paths = find_paths(path, root, |v| match v.get_type() {
         SelectValueType::Array | SelectValueType::Object => v.len().unwrap() > 0,
-        SelectValueType::Long => v.get_long() != 0,
+        SelectValueType::Long => v.get_long().map_or_else(|_| false, |v| v != 0),
```
Comment (MeirShpilraien) [inline src/commands.rs:1941]: Don't we want to clear it here anyway?

File: src/ivalue_manager.rs [hunk 10]
```diff
@@ -219,7 +220,7 @@ impl<'a> IValueKeyHolderWrite<'a> {
         mut op2_fun: F2,
     ) -> Result<Number, RedisError>
     where
-        F1: FnMut(i64, i64) -> i64,
+        F1: FnMut(i64, i64) -> (i64, bool),
```
Comment (MeirShpilraien) [inline src/ivalue_manager.rs:223]: Whats the bool is used for?

File: tests/pytest/test.py [hunk 11]
```diff
@@ -1280,6 +1280,123 @@ def testFilterPrecedence(env):
...
+        for k, v in iter(obj.items()):
+            r.assertTrue(True, message='{} GET {}={}'.format(name, k, v))
+            res = r.execute_command('JSON.GET', name, '$.{}'.format(k))
+            r.assertAlmostEqual(json.loads(res)[0], v, epsilon, message=res)
+            try:
+                # Check no crash on overflow
+                res = r.execute_command('JSON.NUMINCRBY', name, '$.{}'.format(k), '1')
+                if check_num_incr:
+                    r.assertAlmostEqual(json.loads(res)[0], v+1, epsilon, message=res)
+                r.execute_command('JSON.NUMINCRBY', name, '$.{}'.format(k), '9223372036854775808')
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:1348]: I am not follow, if one of the checks raise error we are not checking the rest? What am I missing?

File: tests/pytest/test.py [hunk 12]
```diff
@@ -1280,6 +1280,123 @@ def testFilterPrecedence(env):
...
+        res = r.execute_command('JSON.RESP', 'arr')
+        for res_i, arr_i in zip(res[1:], arr):
+            if type(res_i) is int:
+                # Integers greater then max i64 are returned as max i64 by RESP
+                r.assertAlmostEqual(res_i, arr_i if arr_i <= sys.maxsize else sys.maxsize, epsilon, message=res)
+            elif type(res_i) is str:
+                # Double are returned as strings
+                r.assertAlmostEqual(float(res_i), arr_i, epsilon, message=res)
+        for k, v in iter(obj.items()):            
+            r.execute_command('JSON.CLEAR', name, '$.{}'.format(k))
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:1383]: Check result?

File: tests/pytest/test.py [hunk 13]
```diff
@@ -1280,6 +1280,123 @@ def testFilterPrecedence(env):
...
+    check_object_values(doc_float_ok, 'doc_float_ok', sys.float_info.epsilon)
+    check_object_values(doc_float_ok_overflow_mult, 'doc_float_ok_overflow_mult', 0, True, False)
+
+    check_object_values(doc_float_ok_2_overflow_mult, 'doc_float_ok_2_overflow_mult', sys.float_info.epsilon, True, False)
+    
+    # Not using json.dumps with out-of-range values here (would be converted to a string representation such as 'Infinity')
+    r.expect('JSON.SET', 'doc_bad_values', '$', '{}').ok()
+    for k, v in iter(doc_bad_values.items()):
+        r.assertTrue(True, message='SET {}={}'.format(k, v))
+        r.expect('JSON.SET', 'doc_bad_values', '$.{}'.format(k), '{}'.format(v)).error()
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:1398]:
```suggestion
        r.expect('JSON.SET', 'doc_bad_values', '$.{}'.format(k), '{}'.format(v)).error().contians(<some substring from the expected error>)
```

Comment (oshadmi) [reply tests/pytest/test.py:1398]: This errors are JSON parse errors, such as with these commands
```
"JSON.SET" "doc_bad_values" "$.beyond_max_f64" "inf"
"JSON.SET" "doc_bad_values" "$.below_min_f64" "-inf"
```
these error messages are issued:
`expected value at line 1 column 1`
`invalid number at line 1 column 2`

File: tests/pytest/test.py (hunk 1 above)
Comment (oshadmi) [reply tests/pytest/test.py]: It is not enabled by default
`--debug-print         print debug messages (default: False)`

File: tests/pytest/test.py (hunk 2 above)
Comment (oshadmi) [reply tests/pytest/test.py:1345]: The verification will fail, e.g., the result of NUMINCRBY will not be `v+1`

File: tests/pytest/test.py (hunk 6 above)
Comment (oshadmi) [reply tests/pytest/test.py:1397]: Not enabled by default

File: src/commands.rs (hunk 9 above)
Comment (oshadmi) [reply src/commands.rs:1941]: Yes, fixing it.

File: src/ivalue_manager.rs (hunk 10 above)
Comment (oshadmi) [reply src/ivalue_manager.rs:223]: The bool returns whether there was an overflow.
Using `overflowing_add` and `overflowing_mul` returning the wrapped value if overflowed (should be the same result as with Release builds, only not skipping the overflow check)

File: tests/pytest/test.py (hunk 11 above)
Comment (oshadmi) [reply tests/pytest/test.py:1348]: Right, seemed enough to check first overflow/unsupported value, but we can test all.

File: src/c_api.rs [hunk 14]
```diff
@@ -177,6 +191,19 @@ pub fn json_api_get_double<M: Manager>(_: M, json: *const c_void, val: *mut c_do
             unsafe { *val = json.get_long() as f64 };
             Status::Ok as c_int
         }
+
+        SelectValueType::ULong => {
```
Comment (MeirShpilraien) [inline src/c_api.rs:195]: Why would someone call get_double if the value is ulong? I see we did it for long as well but I wonder if its not an API missuse?

File: src/commands.rs [hunk 15]
```diff
@@ -418,6 +425,7 @@ impl<'a, V: SelectValue + 'a> KeyValue<'a, V> {
             (SelectValueType::Null, SelectValueType::Null) => true,
             (SelectValueType::Bool, SelectValueType::Bool) => a.get_bool() == b.get_bool(),
             (SelectValueType::Long, SelectValueType::Long) => a.get_long() == b.get_long(),
+            (SelectValueType::ULong, SelectValueType::ULong) => a.get_ulong() == b.get_ulong(),
```
Comment (MeirShpilraien) [inline src/commands.rs:428]: So ulong will never be equal to long? I guess this claim also true when comparing long to double, but maybe here it will be more noticeable?

File: src/include/rejson_api.h [hunk 16]
```diff
@@ -20,7 +20,11 @@ typedef enum JSONType {
   JSONType_Object = 4,
   JSONType_Array = 5,
   JSONType_Null = 6,
-  JSONType__EOF
+  JSONType__EOF = 7, // keep this for backward compatibility
```
Comment (MeirShpilraien) [inline src/include/rejson_api.h:23]: What is this value means and where is it used?

File: src/ivalue_manager.rs [hunk 17]
```diff
@@ -382,15 +442,33 @@ impl<'a> WriteHolder<IValue, IValue> for IValueKeyHolderWrite<'a> {
     }
 
     fn incr_by(&mut self, path: Vec<String>, num: &str) -> Result<Number, RedisError> {
-        self.do_num_op(path, num, |i1, i2| i1 + i2, |f1, f2| f1 + f2)
+        self.do_num_op(
+            path,
+            num,
+            |i1, i2| i1.overflowing_add(i2),
```
Comment (MeirShpilraien) [inline src/ivalue_manager.rs:448]: Can this really happened now? If it will overflow we will move to u64 no?

File: src/ivalue_manager.rs [hunk 18]
```diff
@@ -382,15 +442,33 @@ impl<'a> WriteHolder<IValue, IValue> for IValueKeyHolderWrite<'a> {
...
+            |u1, u2| u1.overflowing_add(u2),
+        )
     }
 
     fn mult_by(&mut self, path: Vec<String>, num: &str) -> Result<Number, RedisError> {
-        self.do_num_op(path, num, |i1, i2| i1 * i2, |f1, f2| f1 * f2)
+        self.do_num_op(
+            path,
+            num,
+            |i1, i2| i1.overflowing_mul(i2),
```
Comment (MeirShpilraien) [inline src/ivalue_manager.rs:458]: Same question as above

File: src/ivalue_manager.rs [hunk 19]
```diff
@@ -382,15 +442,33 @@ impl<'a> WriteHolder<IValue, IValue> for IValueKeyHolderWrite<'a> {
...
+            |u1, u2| u1.overflowing_mul(u2),
+        )
     }
 
     fn pow_by(&mut self, path: Vec<String>, num: &str) -> Result<Number, RedisError> {
-        self.do_num_op(path, num, |i1, i2| i1.pow(i2 as u32), f64::powf)
+        self.do_num_op(
+            path,
+            num,
+            |i1, i2| i1.overflowing_pow(i2 as u32),
```
Comment (MeirShpilraien) [inline src/ivalue_manager.rs:468]: Same question as above

File: src/c_api.rs (hunk 14 above)
Comment (oshadmi) [reply src/c_api.rs:195]: Basically this is an optimization, for example, an array of doubles, in which some may be integers, not requiring to consult the type of each, or avoid [getDouble](https://github.com/RediSearch/RediSearch/blob/4387b045b764c7bfc802646dd0b8af52845b56b3/src/json.c#L143) to fail and then re-try as int.

File: tests/pytest/test.py [hunk 20]
```diff
@@ -1280,6 +1280,156 @@ def testFilterPrecedence(env):
...
+        for k, v in iter(obj.items()):
+            r.assertTrue(True, message='{} GET {}={}'.format(name, k, v))
+            res = r.execute_command('JSON.GET', name, '$.{}'.format(k))
+            r.assertAlmostEqual(json.loads(res)[0], v, epsilon, message=res)
+            
+            values = ['9223372036854775808', '18446744073709551615', '18446744073709551616', '-9223372036854775808', '-18446744073709551615', '-18446744073709551616']
+            
+            # Check NUMINCRBY
+            try:
+                r.assertTrue(True, message='{} NUMINCRBY {}: {} + 1'.format(name, k, v))
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:1353]: If this is only for debug prints please use r.printDebug (or debugPrint, I do not remember exactly)..

File: src/include/rejson_api.h (hunk 16 above)
Comment (oshadmi) [reply src/include/rejson_api.h:23]: Used by RediSearch `JSONType`. Marks a non-existing type (maybe it was wrong to introduce it, but now it exists)

File: tests/pytest/test.py [hunk 21]
```diff
@@ -1280,6 +1280,156 @@ def testFilterPrecedence(env):
...
+    r.expect('JSON.SET', 'doc_int_ok_overflow_mult', '$', json.dumps(doc_int_ok_overflow_mult)).ok()
+    r.expect('JSON.SET', 'doc_int_ok_overflow_incr', '$', json.dumps(doc_int_ok_overflow_incr)).ok()
+    r.expect('JSON.SET', 'doc_int_ok_overflow_both', '$', json.dumps(doc_int_ok_overflow_both)).ok()
+
+    r.expect('JSON.SET', 'doc_float_ok', '$', json.dumps(doc_float_ok)).ok()
+    r.expect('JSON.SET', 'doc_float_ok_overflow_incr', '$', json.dumps(doc_float_ok_overflow_incr)).ok()
+
+    r.expect('JSON.SET', 'doc_float_ok_2_overflow_incr', '$', json.dumps(doc_float_ok_2_overflow_incr)).ok()
+    
+    def check_object_values(obj, name, epsilon, check_num_incr=True, check_num_mult=True):
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:1339]: I honestly would prefer to have a test which is easier to read even at a cost of code duplication. It is very hard to follow this test and understand what it is testing. I also prefer a more simple and short tests even at the cost of many test functions.

**Review by MeirShpilraien — COMMENTED** [review]
Few more comments.

File: src/c_api.rs (hunk 14 above)
Comment (MeirShpilraien) [reply src/c_api.rs:195]: It could be done on the user side of the API, I believe API's should be minimal and simple.

File: src/commands.rs (hunk 15 above)
Comment (MeirShpilraien) [reply src/commands.rs:428]: Ohh nice, lets add a comment here that explains it.

File: src/include/rejson_api.h (hunk 16 above)
Comment (MeirShpilraien) [reply src/include/rejson_api.h:23]: Do we ever return such a value? How come we will have none existing type?

File: src/ivalue_manager.rs (hunk 19 above)
Comment (oshadmi) [reply src/ivalue_manager.rs:468]: Although `NUMPOWBY` is deprecated, since it is still here, same behavior as add and mul

File: src/include/rejson_api.h (hunk 16 above)
Comment (oshadmi) [reply src/include/rejson_api.h:23]: We do not return it, but if we would use value `7` for a new type, an older RediSearch module might error.
Maybe it was used as the type for uninitialized value.

File: src/c_api.rs (hunk 14 above)
Comment (oshadmi) [reply src/c_api.rs:195]: For the common use case, checking the type, and asking for the corresponding value is minimal and simple.
For double, it can also do more in case it is requested to (and `ULong` is now aligned with `Long`)
Do you think we should revert this optimization? (could affect RediSearch VecSim)
